File: platform/web/src/lib/camera2d/messages.ts
// Write-side wire builders for the camera2d contract, mirroring the to_wire
// methods in platform/packages/contracts/camera2d/src/wf/contracts/camera2d/
// messages.py. The READ-side interfaces (FrameHeader, StreamParams,
// CameraStatus, GrabReply, Pose) already live in lib/messages.ts; this module
// is the producer half the TS HAL needs.
//
// Float discipline: every field the Python dataclass declares `float` is
// wrapped with f()/farr() so it serializes as CBOR float64 even when its JS
// value is integral (1.0 === 1 in JS). Guaranteed-int fields (t_capture, w, h,
// seq, t, quality, roi) stay bare numbers. See cbor.ts for why.
//
// Key ORDER matches each Python to_wire emission order exactly (cbor2 keeps
// insertion order). The ONLY omitted-when-null key in the whole contract is
// FrameHeader.pose; every other nullable field is emitted as explicit null.

import { type Flt, type Wire, f, farr } from "./cbor";
import type { CameraStatus, Pose, StreamParams } from "../messages";

export type WireMap = { [k: string]: Wire };

export const ENCODING_JPEG = "jpeg";
export const ENCODING_BAYER_RG8 = "BayerRG8";
export const ENCODING_MONO8 = "Mono8";
export const ENCODINGS = [ENCODING_JPEG, ENCODING_BAYER_RG8, ENCODING_MONO8];

/** Realm-less optical frame name `camera2d/{cid}/optical` (keys.py). The
 *  FrameHeader.frame_id every camera frame carries. */
export function opticalFrame(cid: string): string {
  return `camera2d/${cid}/optical`;
}

// ── Ack ──────────────────────────────────────────────────────────────────
export interface Ack {
  ok: boolean;
  error: string | null;
}

export function ackWire(ok: boolean, error: string | null = null): WireMap {
  return { ok, error };
}

// ── FrameSpec / StreamParams (parsed from cmd payloads) ────────────────────
export interface FrameSpec {
  scale: number;
  roi: number[] | null;
  encoding: string;
  quality: number;
}

const FRAME_SPEC_DEFAULTS: FrameSpec = {
  scale: 1.0,
  roi: null,
  encoding: ENCODING_BAYER_RG8,
  quality: 95,
};

/** Parse + validate a cmd/grab FrameSpec from a decoded wire dict, merging the
 *  resource grab_defaults. Throws on invalid input (the error string is
 *  surfaced in the GrabReply, matching the Python __post_init__ messages). */
export function parseFrameSpec(
  d: Record<string, unknown>,
  defaults: Partial<FrameSpec> = {},
): FrameSpec {
  const m = { ...FRAME_SPEC_DEFAULTS, ...defaults, ...d };
  const scale = Number(m.scale);
  const quality = Number(m.quality);
  const encoding = String(m.encoding);
  const roi = m.roi == null ? null : (m.roi as number[]).map((v) => Math.trunc(Number(v)));
  if (!(scale > 0 && scale <= 1)) throw new Error(`scale must be in (0, 1], got ${scale}`);
  if (!ENCODINGS.includes(encoding)) {
    throw new Error(`encoding must be one of ${ENCODINGS.join(",")}, got '${encoding}'`);
  }
  if (!(quality >= 1 && quality <= 100)) {
    throw new Error(`quality must be in [1, 100], got ${quality}`);
  }
  if (roi !== null) {
    if (roi.length !== 4) throw new Error(`roi must be [x, y, w, h], got ${JSON.stringify(roi)}`);
    const [x, y, w, h] = roi;
    if (x < 0 || y < 0) throw new Error(`roi x/y must be non-negative, got ${JSON.stringify(roi)}`);
    if (w <= 0 || h <= 0) throw new Error(`roi w/h must be positive, got ${JSON.stringify(roi)}`);
  }
  if (encoding === ENCODING_BAYER_RG8 && scale !== 1.0) {
    throw new Error("scale requires jpeg encoding");
  }
  return { scale, roi, encoding, quality };
}

/** Parse + validate a cmd/stream_start StreamParams (FrameSpec + rate_hz). */
export function parseStreamParams(
  d: Record<string, unknown>,
  defaults: Partial<StreamParams> = {},
): StreamParams {
  const spec = parseFrameSpec(d, defaults);
  const rate_hz = Number(d.rate_hz ?? defaults.rate_hz ?? 15.0);
  if (!(rate_hz > 0 && rate_hz <= 60)) {
    throw new Error(`rate_hz must be in (0, 60], got ${rate_hz}`);
  }
  return { ...spec, rate_hz };
}

/** StreamParams wire is FLAT: {scale, roi, encoding, quality, rate_hz}. */
export function streamParamsWire(sp: StreamParams): WireMap {
  return {
    scale: f(sp.scale),
    roi: sp.roi === null ? null : sp.roi.map((v) => Math.trunc(v)),
    encoding: sp.encoding,
    quality: Math.trunc(sp.quality),
    rate_hz: f(sp.rate_hz),
  };
}

// ── ConfigureCmd (parsed from cmd/configure) ───────────────────────────────
export interface ConfigureCmd {
  exposure_us: number | null;
  gain_db: number | null;
  auto_exposure: boolean | null;
  auto_gain: boolean | null;
  auto_wb: boolean | null;
  wb_red: number | null;
  wb_blue: number | null;
}

export function parseConfigureCmd(d: Record<string, unknown>): ConfigureCmd {
  const num = (v: unknown): number | null => (v == null ? null : Number(v));
  const bool = (v: unknown): boolean | null => (v == null ? null : Boolean(v));
  return {
    exposure_us: num(d.exposure_us),
    gain_db: num(d.gain_db),
    auto_exposure: bool(d.auto_exposure),
    auto_gain: bool(d.auto_gain),
    auto_wb: bool(d.auto_wb),
    wb_red: num(d.wb_red),
    wb_blue: num(d.wb_blue),
  };
}

// ── pose ───────────────────────────────────────────────────────────────────
function poseWire(pose: Pose): WireMap {
  return { frame: pose.frame, xyz: farr(pose.xyz), quat: farr(pose.quat) };
}

// ── FrameHeader ────────────────────────────────────────────────────────────
export interface FrameHeaderInit {
  t_capture: number | bigint;
  frame_id: string;
  w: number;
  h: number;
  encoding: string;
  exposure_us: number;
  gain_db: number;
  seq: number;
  clock_domain: string;
  pose: Pose | null;
}

/** FrameHeader.to_wire: pose key APPENDED only when present (the one
 *  omitted-when-null field in the contract). */
export function frameHeaderWire(hdr: FrameHeaderInit): WireMap {
  const d: WireMap = {
    t_capture: hdr.t_capture,
    frame_id: hdr.frame_id,
    w: Math.trunc(hdr.w),
    h: Math.trunc(hdr.h),
    encoding: hdr.encoding,
    exposure_us: f(hdr.exposure_us),
    gain_db: f(hdr.gain_db),
    seq: Math.trunc(hdr.seq),
    clock_domain: hdr.clock_domain,
  };
  if (hdr.pose !== null) d.pose = poseWire(hdr.pose);
  return d;
}

// ── GrabReply ──────────────────────────────────────────────────────────────
export function grabReplyWire(
  ok: boolean,
  error: string | null,
  header: WireMap | null,
  data: Uint8Array | null,
): WireMap {
  return { ok, error, header, data };
}

// ── CameraStatus ───────────────────────────────────────────────────────────
export function cameraStatusWire(s: CameraStatus): WireMap {
  const flt = (v: number | null): Flt | null => (v === null ? null : f(v));
  return {
    t: s.t,
    connected: s.connected,
    streaming: s.streaming,
    stream: s.stream === null ? null : streamParamsWire(s.stream),
    exposure_us: flt(s.exposure_us),
    gain_db: flt(s.gain_db),
    achieved_rate_hz: f(s.achieved_rate_hz),
    error: s.error,
  };
}

File: platform/packages/contracts/camera2d/src/wf/contracts/camera2d/messages.py
"""Hand-written `camera2d` contract wire messages (no codegen in phase 1).

Wire field names in ``to_wire``/``from_wire`` are normative. Conventions:
timestamps int nanoseconds, exposure microseconds, gain dB. Frame payloads
travel OUTSIDE these messages: zenoh payload = image bytes, zenoh
attachment = CBOR-encoded :class:`FrameHeader`.
"""

from __future__ import annotations

from dataclasses import dataclass

from wf.core.time import CLOCK_HOST

ENCODING_BAYER_RG8 = "BayerRG8"
ENCODING_JPEG = "jpeg"
ENCODING_MONO8 = "Mono8"
ENCODINGS = (ENCODING_BAYER_RG8, ENCODING_JPEG, ENCODING_MONO8)


@dataclass
class FrameHeader:
    """CBOR attachment on every frame topic; embedded in GrabReply.

    The bus-wide frame convention: producers publish image bytes as the payload
    and this header as the attachment on an ``.../image`` topic. Derived frame
    producers preserve ``t_capture`` and ``frame_id``, assign their own ``seq``,
    and update ``w``/``h``/``encoding``.
    Origin is carried by the topic; there is no ``source`` field.
    """

    t_capture: int  # ns, exposure midpoint
    frame_id: str  # camera2d/{cid}/optical
    w: int
    h: int
    encoding: str  # ENCODINGS member
    exposure_us: float
    gain_db: float
    seq: int  # monotonic per producer process (shared stream+grab counter)
    clock_domain: str = CLOCK_HOST  # CLOCK_CAMERA once timesync calibrated
    # World<-optical camera pose AT CAPTURE, when known: {frame, xyz[3] (m),
    # quat[4] [x,y,z,w] scalar-last}. ``frame`` is the base/world frame the pose
    # is expressed in (== ``frame_id``'s parent, ``arm/{rid}/base`` in v0).
    # ``None`` when the producer has no pose (e.g. an eye-in-hand camera before
    # its arm's flange streams, or the conformance suite). OPTIONAL/additive —
    # older frames decode with ``pose=None``; consumers ignore it freely.
    pose: dict | None = None

    def __post_init__(self):
        if self.encoding not in ENCODINGS:
            raise ValueError(
                f"encoding must be one of {ENCODINGS}, got {self.encoding!r}"
            )

    def to_wire(self) -> dict:
        d = {
            "t_capture": int(self.t_capture),
            "frame_id": self.frame_id,
            "w": int(self.w),
            "h": int(self.h),
            "encoding": self.encoding,
            "exposure_us": float(self.exposure_us),
            "gain_db": float(self.gain_db),
            "seq": int(self.seq),
            "clock_domain": self.clock_domain,
        }
        if self.pose is not None:
            d["pose"] = self.pose
        return d

    @classmethod
    def from_wire(cls, d: dict) -> "FrameHeader":
        return cls(
            t_capture=d["t_capture"],
            frame_id=d["frame_id"],
            w=d["w"],
            h=d["h"],
            encoding=d["encoding"],
            exposure_us=d["exposure_us"],
            gain_db=d["gain_db"],
            seq=d["seq"],
            clock_domain=d.get("clock_domain", CLOCK_HOST),
            pose=d.get("pose"),
        )

@dataclass
class ProducerGrant:
    """Current browser producer grant, including restart-safe fencing."""

    client_id: str
    user: str
    authority_id: str
    epoch: int
    granted_at: int
    expires_at: int

    def to_wire(self) -> dict:
        return {
            "client_id": self.client_id,
            "user": self.user,
            "authority_id": self.authority_id,
            "epoch": int(self.epoch),
            "granted_at": int(self.granted_at),
            "expires_at": int(self.expires_at),
        }

    @classmethod
    def from_wire(cls, d: dict) -> "ProducerGrant":
        return cls(
            client_id=d["client_id"],
            user=d["user"],
            authority_id=d["authority_id"],
            epoch=d["epoch"],
            granted_at=d["granted_at"],
            expires_at=d["expires_at"],
        )


@dataclass
class ProducerFrame:
    """CBOR attachment on a candidate frame published to producer ingress."""

    client_id: str
    authority_id: str
    epoch: int
    captured_at: int
    w: int
    h: int
    encoding: str
    exposure_us: float
    gain_db: float
    pose: dict | None = None

    def __post_init__(self):
        if self.encoding != ENCODING_JPEG:
            raise ValueError("browser producer frames must use jpeg encoding")
        if self.w <= 0 or self.h <= 0:
            raise ValueError("producer frame dimensions must be positive")

    def to_wire(self) -> dict:
        d = {
            "client_id": self.client_id,
            "authority_id": self.authority_id,
            "epoch": int(self.epoch),
            "captured_at": int(self.captured_at),
            "w": int(self.w),
            "h": int(self.h),
            "encoding": self.encoding,
            "exposure_us": float(self.exposure_us),
            "gain_db": float(self.gain_db),
        }
        if self.pose is not None:
            d["pose"] = self.pose
        return d

    @classmethod
    def from_wire(cls, d: dict) -> "ProducerFrame":
        return cls(
            client_id=d["client_id"],
            authority_id=d["authority_id"],
            epoch=d["epoch"],
            captured_at=d["captured_at"],
            w=d["w"],
            h=d["h"],
            encoding=d["encoding"],
            exposure_us=d["exposure_us"],
            gain_db=d["gain_db"],
            pose=d.get("pose"),
        )



@dataclass
class FrameSpec:
    """``cmd/grab`` payload; the shared frame-processing parameters.

    Grab and stream share this exact parameter surface — a
    :class:`StreamParams` is a FrameSpec plus ``rate_hz``. ``roi`` is a
    software crop ``[x, y, w, h]`` in full-resolution pixel coordinates
    (the driver rounds raw-output ROIs down to even values to preserve the
    Bayer CFA phase).
    """

    scale: float = 1.0  # resize factor after crop; jpeg only; 1.0 = native
    roi: list[int] | None = None  # [x, y, w, h] full-res px; None = full
    encoding: str = ENCODING_BAYER_RG8
    quality: int = 95  # JPEG only; ignored for raw

    def __post_init__(self):
        if not 0 < self.scale <= 1:
            raise ValueError(f"scale must be in (0, 1], got {self.scale}")
        if self.encoding not in ENCODINGS:
            raise ValueError(
                f"encoding must be one of {ENCODINGS}, got {self.encoding!r}"
            )
        if not 1 <= self.quality <= 100:
            raise ValueError(f"quality must be in [1, 100], got {self.quality}")
        if self.roi is not None:
            if len(self.roi) != 4:
                raise ValueError(f"roi must be [x, y, w, h], got {self.roi!r}")
            x, y, w, h = self.roi
            if x < 0 or y < 0:
                raise ValueError(f"roi x/y must be non-negative, got {self.roi!r}")
            if w <= 0 or h <= 0:
                raise ValueError(f"roi w/h must be positive, got {self.roi!r}")
        if self.encoding == ENCODING_BAYER_RG8 and self.scale != 1.0:
            raise ValueError("scale requires jpeg encoding")

    def to_wire(self) -> dict:
        return {
            "scale": float(self.scale),
            "roi": None if self.roi is None else [int(v) for v in self.roi],
            "encoding": self.encoding,
            "quality": int(self.quality),
        }

    @classmethod
    def from_wire(cls, d: dict) -> "FrameSpec":
        roi = d.get("roi")
        return cls(
            scale=d.get("scale", 1.0),
            roi=None if roi is None else list(roi),
            encoding=d.get("encoding", ENCODING_BAYER_RG8),
            quality=d.get("quality", 95),
        )


@dataclass
class StreamParams(FrameSpec):
    """``cmd/stream_start`` payload; echoed in :class:`CameraStatus`.

    ``rate_hz`` above the sensor rate (~24 fps at full res, GigE-saturated)
    just publishes every frame — no error.
    """

    rate_hz: float = 15.0  # publish rate; decimated from sensor rate

    def __post_init__(self):
        super().__post_init__()
        if not 0 < self.rate_hz <= 60:
            raise ValueError(f"rate_hz must be in (0, 60], got {self.rate_hz}")

    def to_wire(self) -> dict:
        # Wire shape is FLAT: {rate_hz, scale, roi, encoding, quality}.
        d = super().to_wire()
        d["rate_hz"] = float(self.rate_hz)
        return d

    @classmethod
    def from_wire(cls, d: dict) -> "StreamParams":
        roi = d.get("roi")
        return cls(
            rate_hz=d.get("rate_hz", 15.0),
            scale=d.get("scale", 1.0),
            roi=None if roi is None else list(roi),
            encoding=d.get("encoding", ENCODING_BAYER_RG8),
            quality=d.get("quality", 95),
        )


@dataclass
class ConfigureCmd:
    """``cmd/configure`` payload; all None = no-op.

    Auto toggles map True -> GenICam ``Continuous``, False -> ``Off``.
    """

    exposure_us: float | None = None
    gain_db: float | None = None
    auto_exposure: bool | None = None  # GenICam ExposureAuto Continuous/Off
    auto_gain: bool | None = None
    auto_wb: bool | None = None
    wb_red: float | None = None
    wb_blue: float | None = None

    def to_wire(self) -> dict:
        return {
            "exposure_us": None
            if self.exposure_us is None
            else float(self.exposure_us),
            "gain_db": None if self.gain_db is None else float(self.gain_db),
            "auto_exposure": self.auto_exposure,
            "auto_gain": self.auto_gain,
            "auto_wb": self.auto_wb,
            "wb_red": None if self.wb_red is None else float(self.wb_red),
            "wb_blue": None if self.wb_blue is None else float(self.wb_blue),
        }

    @classmethod
    def from_wire(cls, d: dict) -> "ConfigureCmd":
        return cls(
            exposure_us=d.get("exposure_us"),
            gain_db=d.get("gain_db"),
            auto_exposure=d.get("auto_exposure"),
            auto_gain=d.get("auto_gain"),
            auto_wb=d.get("auto_wb"),
            wb_red=d.get("wb_red"),
            wb_blue=d.get("wb_blue"),
        )


@dataclass
class GrabReply:
    """``cmd/grab`` envelope ``value``: the captured frame — header + image
    bytes (the same frame is also published on the image topic)."""

    header: FrameHeader
    data: bytes

    def to_wire(self) -> dict:
        return {"header": self.header.to_wire(), "data": self.data}

    @classmethod
    def from_wire(cls, d: dict) -> "GrabReply":
        return cls(header=FrameHeader.from_wire(d["header"]), data=d["data"])


@dataclass
class CameraStatus:
    """Camera status keepalive (~1 Hz)."""

    t: int
    connected: bool
    streaming: bool
    stream: StreamParams | None  # active params, None when idle
    exposure_us: float | None  # last read from nodemap; None when disconnected
    gain_db: float | None
    achieved_rate_hz: float  # measured published-frame rate (0 when idle)
    error: str | None = None

    def to_wire(self) -> dict:
        return {
            "t": int(self.t),
            "connected": bool(self.connected),
            "streaming": bool(self.streaming),
            "stream": None if self.stream is None else self.stream.to_wire(),
            "exposure_us": None
            if self.exposure_us is None
            else float(self.exposure_us),
            "gain_db": None if self.gain_db is None else float(self.gain_db),
            "achieved_rate_hz": float(self.achieved_rate_hz),
            "error": self.error,
        }

    @classmethod
    def from_wire(cls, d: dict) -> "CameraStatus":
        stream = d.get("stream")
        return cls(
            t=d["t"],
            connected=d["connected"],
            streaming=d["streaming"],
            stream=None if stream is None else StreamParams.from_wire(stream),
            exposure_us=d.get("exposure_us"),
            gain_db=d.get("gain_db"),
            achieved_rate_hz=d["achieved_rate_hz"],
            error=d.get("error"),
        )


#: Registered envelope error ``reason`` values (wire-contract RFC §5).
ERROR_REASONS = (
    "bad_request",
    "streaming",
    "unsupported_encoding",
    "grab_failed",
    "stream_failed",
    "configure_failed",
    "held_by",
)
